File: JuliaTutorial/runSim.jl
using DelimitedFiles, Distributions, Primes, Random, Statistics
using Distributions
n = parse(Int, ARGS[1])
reps = parse(Int, ARGS[2])
seed = parse(Int, ARGS[3])
if length(ARGS) < 4
    d = Normal()
else
    d = eval(Meta.parse(ARGS[4]))
end

"""
    est_mean_prime(x)

Estimate mean by averaging prime-indexed data in `x`.
"""
function est_mean_prime(x::Vector{<:Real})
    s, c = zero(eltype(x)), 0
    for i in eachindex(x)
        if Primes.isprime(i)
            s += x[i]
            c += 1
        end
    end
    s / c
end

"""
    est_mean_avg(x)

Estimate mean by sample average.
"""
function est_mean_avg(x::Vector{<:Real})
    mean(x)
end

function compare_methods(n::Int, d::ContinuousUnivariateDistribution)
    x = rand(d, n)
    est_mean_avg(x), est_mean_prime(x)
end

# Simulate `reps` replicates of sample size `n` from distribution `d` using seed `s`
simres = zeros(reps, 2)
Random.seed!(seed)
for r in 1:reps
    x = rand(d, n)
    simres[r, 1] = est_mean_avg(x)
    simres[r, 2] = est_mean_prime(x)
end

outfile = "simresults/n_$(n)_reps_$(reps)_dist_$(d).txt"
DelimitedFiles.writedlm(outfile, simres, ",")
# may add an optional zip step here to reduce storage

File: RTutorial/runSim.R
args = commandArgs(trailingOnly=TRUE)
n = as.integer(args[1])
reps = as.integer(args[2])
s = as.integer(args[3])
if(length(args) < 4){
  d = "rnorm(n)"
} else {
  d = args[4]
}
oFile <- paste("simresults/", "n_", n, "d_", d, ".txt", sep="")


## check if a given integer is prime
isPrime = function(n) {
  if (n <= 3) {
    return (TRUE)
  }
  if (any((n %% 2:floor(sqrt(n))) == 0)) {
    return (FALSE)
  }
  return (TRUE)
}

## estimate mean only using observation with prime indices
estMeanPrimes = function (x) {
  n = length(x)
  ind = sapply(1:n, isPrime)
  return (mean(x[ind]))
}

# Simulate `reps` replicates of sample size `n` from distribution `d` using seed `s`
simres = data.frame(est_mean_avg = double(reps), est_mean_prime = double(reps))
set.seed(s)
for (r in 1:reps){
  x = eval(parse(text = d))
  simres[r, 1] = mean(x)
  simres[r, 2] = estMeanPrimes(x)
}

write.csv(simres, file = oFile, row.names = F)



